backlogger.py

```diff
@@ -236,10 +238,38 @@ def render_influxdb(data):
                 print("\n".join(line for line in render_influxdb(data)))
             else:
                 initialize_md(data)
+                all_good, rows, bad_queries = render_table(data)
                 with open("index.md", "a") as md:
-                    all_good, rows = render_table(data)
                     for row in rows:
                         md.write("|".join(row) + "\n")
+                if os.environ.get('STATE_FOLDER'):
+                    old_state_file = os.path.join(os.environ['STATE_FOLDER'], "state.json")
+                    if os.path.exists(old_state_file):
+                        # open state.json from last run, see if anything changed and send slack notification if needed
+                        with open(old_state_file, "r") as sj:
+                            state = json.load(sj)
+                        old_bad_queries = set(state["bad_queries"].keys())
+                        new_bad_queries = set(bad_queries.keys())
+                        fixed_queries = old_bad_queries - new_bad_queries
+                        broken_queries = new_bad_queries - old_bad_queries
+                        msg = None
+                        if broken_queries:
+                            # something new broke
+                            msg = f":red_circle: Some queries are exceeding limits:"
+                            for query in new_bad_queries:
+                                qd = bad_queries[query]
+                                msg += f"\n• {query} (Issue count {qd['issue_count']} exceeding limit of {qd['limits']})"
+                        elif fixed_queries and not new_bad_queries:
+                            # this is the first green run so let's let everyone know
+                            msg = f":green_heart: All queries within limits again!"
+                        if msg and os.environ.get('SLACK_WEBHOOK_URL'):
+                            r = requests.post(os.environ['SLACK_WEBHOOK_URL'], json={"msg": msg})
+                with open("state.json", "w") as sj:
+                    state = {
+                        "bad_queries": bad_queries,
+                        "updated": datetime.now().isoformat()
+                    }
+                    json.dump(state, sj)
```

Comment:
This code should likely be extracted into its own function. Maybe tests could also be extended (although I'm not sure whether it is worth it for a feature like this).

Reply:
Yeah. Maybe `update_state`. That should also make it easy enough to add a unit test that checks 1. if there's any changes 2. if the JSON has useful state in it.